### PR TITLE
feat: payload distributor / demuxer for ingester response

### PR DESCRIPTION
**Pull request by crepererum**
Helper tool to demux the stream that the ingester returns into multiple data streams that will be consumed by DataFusion. See module documentation for details. Will be used for #8350.

### CODE REVIEW

File: querier/src/ingester/v2/payload_distributor.rs [hunk 1]
```diff
@@ -0,0 +1,645 @@
+//! Tooling to distribute a singe payload stream to multiple output streams.
```
Comment (mhilton) [inline querier/src/ingester/v2/payload_distributor.rs:1]: s/singe/single/

File: querier/src/ingester/v2/payload_distributor.rs [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+//!
+//! # Requirements
+//! This is mostly designed to demux the ingester response stream. The requirements are:
+//!
+//! - **Eager Polling:** The response stream should be polled eagerly even when the demuxed streams are NOT consumed yet. This
+//!   is important to get the data of the wire as quickly as possible, to help the ingester with its data managemene, and
```
Comment (mhilton) [inline querier/src/ingester/v2/payload_distributor.rs:7]: s/managemene/management/

File: querier/src/ingester/v2/payload_distributor.rs [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+//!   trusted component though and we want to emphasize _eager polling_, so the buffer should be unbounded.
+//!
+//! # Alternatives
+//! The following alternatives were considered:
+//!
+//! ## tokio broadcast
+//! Use [`tokio::sync::broadcast`]. This broadcasts a data from a single async source to multiple receivers. Could be
+//! combined with [`ReceiverStream`]. However this approach has multiple drawbacks:
+//!
+//! - **Consumer Restarts:** Broadcasts are only send o existing consumers. Receiving old data on new consumer ends is
```
Comment (mhilton) [inline querier/src/ingester/v2/payload_distributor.rs:26]: s/send o/sent to/?

**Review by mhilton — APPROVED** [review]
LGTM. I hope you don't mind me pedantically pointing out the typos.

File: querier/src/ingester/v2/payload_distributor.rs (hunk 1 above)
Comment (crepererum) [reply querier/src/ingester/v2/payload_distributor.rs:1]: done

File: querier/src/ingester/v2/payload_distributor.rs (hunk 2 above)
Comment (crepererum) [reply querier/src/ingester/v2/payload_distributor.rs:7]: done

**Pull request merged**